llvm/lib/CodeGen/SelectionDAG/ScheduleDAGFast.cpp

```diff
@@ -231,6 +231,11 @@ SUnit *ScheduleDAGFast::CopyAndMoveSuccessors(SUnit *SU) {
     if (!TII->unfoldMemoryOperand(*DAG, N, NewNodes))
       return nullptr;
 
+    // unfolding an x86 RMW operation results in store, dec, load which
+    // can't be handled here so quit
+    if (NewNodes.size() == 3)
+      return nullptr;
+
```

Comment:
Should this put in X86 specific function? Otherwise, it affect other targets.

Reply:
[patch](https://github.com/llvm/llvm-project/commit/7c7ba1baa1bb339273ee279083e99b7eddb0bf79) in ScheduleDAGRRList.cpp was fixed in this way, then maybe need to add 1 function to TargetInstrInfo if that.
But I may raise another PR which could disable unfold for X86 RMW instrs, I just find it could also fix it, if that patch could be landed then I would abandon this one otherwise we may add a function to TargetInstrInfo.cpp if we want to use specific function.

Reply:
Why can't add it in `unfoldMemoryOperand` if it is required in each time?

Reply:
If that then it's same to my solution which disable unfold for RMW, and we need to clarify if RMW would not be used every where or just here.(Although this signature for unfoldMemoryOperand just used here, I don't know if it would be used in the future.

Reply:
I don't understand this PR, either. Shouldn't https://github.com/llvm/llvm-project/pull/67288 be enough?

Reply:
Yes, I raised this first, then I find it can also be fixed by https://github.com/llvm/llvm-project/pull/67288. If that patch landed then this patch should be abandoned.